src/rust/engine/fs/src/store.rs

```diff
@@ -279,13 +276,36 @@ impl Store {
         expanded_digests
       })
       .and_then(move |digests| {
+        if digests.len() < 3 {
```

Comment:
Maybe extract the length/size check into a function called something like `upload_is_faster_than_checking_whether_to_upload` so that this code reads:
```
if upload_is_faster_than_checking_whether_to_upload(&digests) {
  return Ok(digests.keys().map(|key| key.clone()).collect::<HashSet<Digest>());
}
```

in a nice self-documenting way?

(And give that method a little comment along the lines of `// These values are guesses, feel free to tweak them`)